### PR TITLE
Replace NotNullAttribute with thrown exceptions

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

**Comment by kirthik** [comment]
:shipit:

File: src/Microsoft.AspNet.Localization/CookieRequestCultureProvider.cs [hunk 1]
```diff
@@ -66,8 +75,13 @@ public static string MakeCookieValue([NotNull] RequestCulture requestCulture)
         /// </summary>
         /// <param name="value">The cookie value to parse.</param>
         /// <returns>The <see cref="RequestCulture"/> or <c>null</c> if parsing fails.</returns>
-        public static RequestCulture ParseCookieValue([NotNull] string value)
+        public static RequestCulture ParseCookieValue(string value)
         {
+            if (value == null)
```
Comment (hishamco) [inline src/Microsoft.AspNet.Localization/CookieRequestCultureProvider.cs]: ```if(string.NullOrEmpty(value))```

File: src/Microsoft.Extensions.Localization/ResourceManagerWithCultureStringLocalizer.cs [hunk 2]
```diff
@@ -26,31 +27,66 @@ public class ResourceManagerWithCultureStringLocalizer : ResourceManagerStringLo
...
             }
         }
 
         /// <inheritdoc />
-        public override LocalizedString this[[NotNull] string name, params object[] arguments]
+        public override LocalizedString this[string name, params object[] arguments]
         {
             get
             {
+                if (name == null)
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization/ResourceManagerWithCultureStringLocalizer.cs]: ```csharp if(string.NullOrEmpty(name))```

File: src/Microsoft.Extensions.Localization/ResourceManagerWithCultureStringLocalizer.cs [hunk 3]
```diff
@@ -26,31 +27,66 @@ public class ResourceManagerWithCultureStringLocalizer : ResourceManagerStringLo
...
             _culture = culture;
         }
 
         /// <inheritdoc />
-        public override LocalizedString this[[NotNull] string name]
+        public override LocalizedString this[string name]
         {
             get
             {
+                if (name == null)
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization/ResourceManagerWithCultureStringLocalizer.cs]: ```csharp if(string.NullOrEmpty(value))```

File: src/Microsoft.AspNet.Localization/CookieRequestCultureProvider.cs (hunk 1 above)
Comment (hishamco) [inline src/Microsoft.AspNet.Localization/CookieRequestCultureProvider.cs]: ```csharp if(string.NullOrEmpty(value))```

File: src/Microsoft.Extensions.Localization.Abstractions/LocalizedString.cs [hunk 4]
```diff
@@ -27,14 +26,24 @@ public LocalizedString([NotNull] string name, [NotNull] string value)
         /// <param name="name">The name of the string in the resource it was loaded from.</param>
         /// <param name="value">The actual string.</param>
         /// <param name="resourceNotFound">Whether the string was found in a resource. Set this to <c>false</c> to indicate an alternate string value was used.</param>
-        public LocalizedString([NotNull] string name, [NotNull] string value, bool resourceNotFound)
+        public LocalizedString(string name, string value, bool resourceNotFound)
         {
+            if (name == null)
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization.Abstractions/LocalizedString.cs]: ```csharp if(string.NullOrEmpty(name))```

File: src/Microsoft.Extensions.Localization.Abstractions/StringLocalizerExtensions.cs [hunk 5]
```diff
@@ -15,8 +15,21 @@ public static class StringLocalizerExtensions
...
-            [NotNull] string name) => stringLocalizer[name];
+            this IStringLocalizer stringLocalizer,
+            string name)
+        {
+            if (stringLocalizer == null)
+            {
+                throw new ArgumentNullException(nameof(stringLocalizer));
+            }
+
+            if (name == null)
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization.Abstractions/StringLocalizerExtensions.cs]: ```csharp if(string.NullOrEmpty(name))```

File: src/Microsoft.Extensions.Localization.Abstractions/StringLocalizerExtensions.cs [hunk 6]
```diff
@@ -26,16 +39,36 @@ public static class StringLocalizerExtensions
...
+            this IStringLocalizer stringLocalizer,
+            string name,
+            params object[] arguments)
+        {
+            if (stringLocalizer == null)
+            {
+                throw new ArgumentNullException(nameof(stringLocalizer));
+            }
+
+            if (name == null)
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization.Abstractions/StringLocalizerExtensions.cs]: ```csharp if(string.NullOrEmpty(name))```

File: src/Microsoft.Extensions.Localization.Abstractions/StringLocalizerOfT.cs [hunk 7]
```diff
@@ -28,11 +33,32 @@ public StringLocalizer([NotNull] IStringLocalizerFactory factory)
         public virtual IStringLocalizer WithCulture(CultureInfo culture) => _localizer.WithCulture(culture);
 
         /// <inheritdoc />
-        public virtual LocalizedString this[[NotNull] string name] => _localizer[name];
+        public virtual LocalizedString this[string name]
+        {
+            get
+            {
+                if (name == null)
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization.Abstractions/StringLocalizerOfT.cs]: ```csharp if(string.NullOrEmpty(name))```

File: src/Microsoft.Extensions.Localization.Abstractions/StringLocalizerOfT.cs [hunk 8]
```diff
@@ -28,11 +33,32 @@ public StringLocalizer([NotNull] IStringLocalizerFactory factory)
...
+        }
 
         /// <inheritdoc />
-        public virtual LocalizedString this[[NotNull] string name, params object[] arguments] =>
-            _localizer[name, arguments];
+        public virtual LocalizedString this[string name, params object[] arguments]
+        {
+            get
+            {
+                if (name == null)
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization.Abstractions/StringLocalizerOfT.cs]: ```csharp if(string.NullOrEmpty(name))```

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs [hunk 9]
```diff
@@ -34,24 +35,47 @@ public class ResourceManagerStringLocalizer : IStringLocalizer
...
+            {
+                throw new ArgumentNullException(nameof(resourceManager));
+            }
+
+            if (resourceAssemblyWrapper == null)
+            {
+                throw new ArgumentNullException(nameof(resourceAssemblyWrapper));
+            }
+
+            if (baseName == null)
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs]: ```csharp if(string.NullOrEmpty(baseName))```

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs [hunk 10]
```diff
@@ -59,20 +83,30 @@ public class ResourceManagerStringLocalizer : IStringLocalizer
         }
 
         /// <inheritdoc />
-        public virtual LocalizedString this[[NotNull] string name]
+        public virtual LocalizedString this[string name]
         {
             get
             {
+                if (name == null)
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs]: ```csharp if(string.NullOrEmpty(name))```

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs [hunk 11]
```diff
@@ -59,20 +83,30 @@ public class ResourceManagerStringLocalizer : IStringLocalizer
...
             }
         }
 
         /// <inheritdoc />
-        public virtual LocalizedString this[[NotNull] string name, params object[] arguments]
+        public virtual LocalizedString this[string name, params object[] arguments]
         {
             get
             {
+                if (name == null)
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs]: ```csharp if(string.NullOrEmpty(name))```

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs [hunk 12]
```diff
@@ -130,8 +169,13 @@ protected IEnumerable<LocalizedString> GetAllStrings(bool includeAncestorCulture
         /// <param name="name">The name of the string resource.</param>
         /// <param name="culture">The <see cref="CultureInfo"/> to get the string for.</param>
         /// <returns>The resource string, or <c>null</c> if none was found.</returns>
-        protected string GetStringSafely([NotNull] string name, CultureInfo culture)
+        protected string GetStringSafely(string name, CultureInfo culture)
         {
+            if (name == null)
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs]: ```csharp if(string.NullOrEmpty(name))```

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs [hunk 13]
```diff
@@ -65,8 +79,13 @@ public IStringLocalizer Create([NotNull] Type resourceSource)
         /// <param name="baseName">The base name of the resource to load strings from.</param>
         /// <param name="location">The location to load resources from.</param>
         /// <returns>The <see cref="ResourceManagerStringLocalizer"/>.</returns>
-        public IStringLocalizer Create([NotNull] string baseName, string location)
+        public IStringLocalizer Create(string baseName, string location)
         {
+            if (baseName == null)
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: ```csharp if(string.NullOrEmpty(baseName))```

File: src/Microsoft.Extensions.Localization/ResourceManagerWithCultureStringLocalizer.cs [hunk 14]
```diff
@@ -26,31 +27,66 @@ public class ResourceManagerWithCultureStringLocalizer : ResourceManagerStringLo
...
+            {
+                throw new ArgumentNullException(nameof(resourceAssembly));
+            }
+
+            if (baseName == null)
+            {
+                throw new ArgumentNullException(nameof(baseName));
+            }
+
+            if (resourceNamesCache == null)
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization/ResourceManagerWithCultureStringLocalizer.cs]: ```csharp if(string.NullOrEmpty(baseName))```

File: src/Microsoft.Extensions.Localization.Abstractions/LocalizedString.cs (hunk 4 above)
Comment (pranavkm) [inline src/Microsoft.Extensions.Localization.Abstractions/LocalizedString.cs]: I don't plan on changing the semantics of the current code. `NotNull` does not imply null or empty. If that was the intent, the original code would've reflected this behavior.

**Pull request closed without merging**